elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/Schema.java

```diff
@@ -0,0 +1,633 @@
+     */
+    private List<Class<? extends Aggregation>> extractNonDefaultAggregations(
+            String fieldName,
+            Class<?> containingCls
+    ) {
+        if (!isMetricField(fieldName, containingCls)) {
+            String message = String.format(NON_METRIC_ERROR_FORMAT, fieldName);
+            log.error(message);
+            throw new IllegalArgumentException(message);
+        }
+
+        return Arrays.stream(
+                getEntityDictionary().getAttributeOrRelationAnnotation(
+                        containingCls,
+                        MetricAggregation.class,
+                        fieldName
+                ).aggregations()
+        )
+                .collect(
+                        Collectors.collectingAndThen(
+                                Collectors.toList(),
+                                Collections::unmodifiableList
+                        )
+                );
+    }
+
+    /**
+     * Returns the <b>expanded</b> JPQL metric expression, with {@link #TABLE_PREFIX table prefix placeholder}
+     * attached, for computing a specified entity field.
+     * <p>
+     * See {@link #TABLE_PREFIX} for more details.
+     *
+     * @param expandingField  The entity field
+     * @param containingCls  The entity that contains the {@code expandingField}
+     * @param visited  A set of already-visited expression variables, which should be skipped during expansion.
+     *
+     * @return a JPQL expression with placeholder
+     */
+    private String getExpandedMetricExpression(
+            String expandingField,
+            Class<?> containingCls,
+            Set<String> visited
+    ) {
+        if (!getEntityDictionary().isValidField(containingCls, expandingField)) {
+            // hit a non-existing field
+            String message = String.format(
+                    "'%s' not found in '%s'",
+                    expandingField,
+                    containingCls.getName()
+            );
+            log.error(message);
+            throw new IllegalStateException(message);
+        }
+        if (!isMetricField(expandingField, containingCls)) {
+            // hit a non-metric field.
+            String message = String.format(
+                    "'%s' in '%s' is not a defined as a metric",
+                    expandingField,
+                    containingCls.getName()
+            );
+            log.error(message);
+            throw new IllegalStateException(message);
+        }
+
+        if (isBaseMetric(expandingField, containingCls)) {
+            // hit a non-computed metric field
+            return String.format(
+                    "%s.%s",
+                    TABLE_PREFIX,
+                    getMappedColumn(expandingField, containingCls)
+            );
+        }
+
+        // hit a computed metric field if valid. Expand
+        visited.add(expandingField);
+        String metricExpression = getEntityDictionary().getAttributeOrRelationAnnotation(
```

Comment:
Maybe this could come from the schema.

Reply:
This method has been moved to ComputedMetric class, if you are ok with it